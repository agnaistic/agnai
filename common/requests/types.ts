import type { GenerateRequestV2 } from '/srv/adapter/type'

export type PayloadOpts = GenerateRequestV2 & { prompt: string }
