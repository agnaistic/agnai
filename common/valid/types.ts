export type FromTupleLiteral<T> = T extends [...string[], null] | readonly [...string[], null]
  ? Exclude<T[number], null> | undefined
  : T extends [...string[]] | readonly [...string[]]
  ? T[number]
  : never

export type OptionalToPrimitive<T extends OptionalPrimitive> = T extends 'string?'
  ? string
  : T extends 'number?'
  ? number
  : T extends 'boolean?'
  ? boolean
  : never
export type OptionalPrimitive = 'string?' | 'number?' | 'boolean?' | 'any?' | 'unknown?'
export type Primitive = 'string' | 'number' | 'boolean' | 'any' | 'unknown'

export type Reference =
  | OptionalPrimitive
  | Primitive
  | readonly [Primitive]
  | [Primitive]
  | readonly [Validator]
  | readonly [Validator, '?']
  | [Validator]
  | [Validator, '?']
  | Validator
  | (Validator & { '?': any })
  | [...string[]]
  | readonly [...string[]]
  | [...string[], null]
  | readonly [...string[], null]

export type Validator = { [key: string]: Reference } | { readonly [key: string]: Reference }

export type FromOptional<T extends OptionalPrimitive> = T extends 'string?'
  ? string | undefined
  : T extends 'boolean?'
  ? boolean | undefined
  : T extends 'number?'
  ? number | undefined
  : T extends 'any?'
  ? any | undefined
  : T extends 'unknown?'
  ? unknown | undefined
  : never

export type FromPrimitve<T extends Primitive> = T extends 'string'
  ? string
  : T extends 'boolean'
  ? boolean
  : T extends 'number'
  ? number
  : T extends 'any'
  ? any
  : T extends 'unknown'
  ? unknown
  : never

export type FromTuple<T> = T extends
  | [infer U]
  | readonly [infer U]
  | [infer U, '?']
  | readonly [infer U, '?']
  ? U extends Primitive
    ? T extends [U, '?'] | readonly [U, '?']
      ? Array<FromPrimitve<U>> | undefined
      : Array<FromPrimitve<U>>
    : never
  : never

export type FromOptionalTuple<T> = T extends [infer U] | readonly [infer U]
  ? U extends OptionalPrimitive
    ? Array<OptionalToPrimitive<U>> | undefined
    : never
  : never

export type FromTupleBody<T> = T extends [infer U, infer O]
  ? U extends Validator
    ? O extends '?'
      ? Array<UnwrapBody<U>> | undefined
      : Array<UnwrapBody<U>>
    : never
  : T extends readonly [infer U]
  ? U extends Validator
    ? Array<UnwrapBody<U>>
    : never
  : never

export type UnwrapBody<T extends Validator> = {
  -readonly [key in keyof T]: key extends '?'
    ? never
    : T[key] extends Primitive
    ? FromPrimitve<T[key]>
    : T[key] extends OptionalPrimitive
    ? FromOptional<T[key]>
    : T[key] extends [Primitive, '?'] | readonly [Primitive, '?']
    ? FromTuple<T[key]>
    : T[key] extends [Primitive] | readonly [Primitive]
    ? FromTuple<T[key]>
    : T[key] extends [OptionalPrimitive] | readonly [OptionalPrimitive]
    ? FromOptionalTuple<T[key]>
    : T[key] extends [Validator, '?'] | readonly [Validator, '?']
    ? Array<UnwrapBody<T[key][0]>> | undefined
    : T[key] extends [Validator] | readonly [Validator]
    ? FromTupleBody<T[key]>
    : T[key] extends Validator & { '?': any }
    ? Omit<UnwrapBody<T[key]>, '?'> | undefined
    : T[key] extends Validator
    ? UnwrapBody<T[key]>
    : FromTupleLiteral<T[key]>
}
