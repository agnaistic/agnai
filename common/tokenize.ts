export type TokenCounter = (text: string) => number

export type Encoder = {
  count: (text: string) => number
  encode: (text: string) => number[]
  decode: (tokens: number[]) => string
}

export type Tokenizer = {
  decode: (tokens: Int32Array) => string
  encode: (value: string) => Int32Array
}

export async function encode(text: string) {
  const encoder = await import('gpt-3-encoder').then((mod) => mod.encode)
  return encoder(text)
}

export async function tokenize(text: string) {
  const encoder = await getEncoder()
  return encoder(text)
}

export async function decode(tokens: number[]) {
  const decoder = await getDecoder()
  return decoder(tokens)
}

export async function getEncoder() {
  const encoder = await import('gpt-3-encoder').then((mod) => mod.encode)
  return (text: string) => encoder(text).length
}

export async function getDecoder() {
  const decode = await import('gpt-3-encoder').then((mod) => mod.decode)
  return (tokens: number[]) => decode(tokens)
}
