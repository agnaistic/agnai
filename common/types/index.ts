import * as UI from './ui'
import * as Sprite from './sprite'
import * as Memory from './memory'

export * from './schema'
export * from './texttospeech-schema'

export { UI, Sprite, Memory }
