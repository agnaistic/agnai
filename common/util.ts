import { AppSchema } from './types/schema'

export function replace<T extends { _id: string }>(id: string, list: T[], item: Partial<T>) {
  return list.map((li) => (li._id === id ? { ...li, ...item } : li))
}

export function findOne<T extends { _id: string }>(id: string, list: T[]): T | void {
  for (const item of list) {
    if (item._id === id) return item
  }
}

export function toArray<T>(values?: T | T[]): T[] {
  if (values === undefined) return []
  if (Array.isArray(values)) return values
  return [values]
}

export function distinct<T extends { _id: string }>(values: T[]) {
  const set = new Set<string>()
  const next: T[] = []

  for (const item of values) {
    if (set.has(item._id)) continue
    next.push(item)
    set.add(item._id)
  }

  return next
}
export function wait(secs: number) {
  return new Promise((res) => setTimeout(res, secs * 1000))
}

export function waitMs(ms: number) {
  return new Promise((res) => setTimeout(res, ms))
}

export function isObject(val: unknown) {
  return Object.prototype.toString.call(val) === '[object Object]'
}

export function cycleArray<T>(arr: T[], shiftAmount: number) {
  const effectiveShift = shiftAmount % arr.length
  return [...arr.slice(effectiveShift), ...arr.slice(0, effectiveShift)]
}

export const dateStringFromUnix = (unixTimestamp: number): string =>
  new Date(unixTimestamp).toISOString()

export function uniqBy<T, U>(arr: T[], accessor: (el: T) => U) {
  return arr.flatMap((el, i) => (arr.slice(0, i).map(accessor).includes(accessor(el)) ? [] : [el]))
}

export function toDuration(valueSecs: number, full?: boolean) {
  if (valueSecs === 0) {
    if (full) return '0s'
    return '0 seconds'
  }
  const {
    duration: [days, hours, minutes, seconds],
  } = toRawDuration(valueSecs)

  if (full) {
    return [`${days}d`, `${hours}h`, `${minutes}m`, `${seconds}s`]
      .filter((time) => !time.startsWith('0'))
      .join(':')
  }

  if (days) {
    return `${days} ${days > 1 ? 'days' : 'day'}`
  }

  if (hours) {
    return `${hours} ${hours > 1 ? 'hours' : 'hour'}`
  }

  if (minutes) {
    return `${minutes} ${minutes > 1 ? 'mins' : 'min'}`
  }

  return `${seconds} ${seconds > 1 ? 'seconds' : 'second'}`
}

export function elapsedSince(date: string | Date, offsetMs: number = 0) {
  if (!date) return 'unknown'
  const diff = Date.now() - new Date(date).valueOf() + offsetMs
  const elapsed = Math.floor(diff / 1000)
  if (elapsed < 60) return 'a moment'
  return toDuration(Math.floor(elapsed))
}

const ONE_HOUR = 3600
const ONE_DAY = 86400

type Days = number
type Hours = number
type Minutes = number
type Seconds = number

type Duration = [Days, Hours, Minutes, Seconds]

function toRawDuration(valueSecs: number) {
  const days = Math.floor(valueSecs / ONE_DAY)
  const hours = Math.floor(valueSecs / ONE_HOUR) % 24
  const mins = Math.floor(valueSecs / 60) % 60
  const secs = Math.ceil(valueSecs % 60)

  return {
    duration: [days, hours, mins, secs] as Duration,
    seconds: valueSecs,
    text: valueSecs <= 0 ? 'now' : `${days}d:${hours}h:${mins}m:${secs}s`.replace('0d:', ''),
  }
}

/**
 * Remove leading tabs and whitespace on each line from template strings
 * Also 'trim' the text before returning
 */
export function neat(params: TemplateStringsArray, ...rest: string[]) {
  let str = ''
  for (let i = 0; i < params.length; i++) {
    str += params[i] + (rest[i] || '')
  }

  return str
    .split('\n')
    .map((line) => line.replace(/^[\t ]+/g, ''))
    .join('\n')
    .trim()
}

const end = `."'*!?)}]\``.split('')
export function trimSentence(text: string) {
  const last = end.reduce((last, char) => {
    const index = text.lastIndexOf(char)
    return index > last ? index : last
  }, -1)

  if (last === -1) return text
  return text.slice(0, last + 1)
}

export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[^\w\s-]/g, '')
    .replace(/[\s_-]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

// https://stackoverflow.com/a/3561711
export function escapeRegex(string: string) {
  return string.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
}

export function getBotName(
  chat: AppSchema.Chat,
  msg: AppSchema.ChatMessage,
  chars: Record<string, AppSchema.Character>,
  replyAs: AppSchema.Character,
  main: AppSchema.Character,
  impersonate?: AppSchema.Character
) {
  if (!msg.characterId) return replyAs?.name || main.name
  if (msg.characterId.startsWith('temp-')) {
    const temp = chat.tempCharacters?.[msg.characterId]
    if (!temp) return main.name
    return temp.name
  }

  const char =
    msg.characterId && impersonate?._id === msg.characterId ? impersonate : chars[msg.characterId]

  if (!char) {
    return main.name
  }

  return char.name
}

export function eventGenerator<T = any>() {
  const queue: Array<Promise<any>> = []
  let done = false
  let signal = false

  const stream = (async function* () {
    while (!done) {
      await waitMs(1)
      const promise = queue.shift()
      if (!promise) {
        if (signal) {
          done = true
        }
        continue
      }

      const result = await promise
      yield result
    }
  })() as AsyncGenerator<T, T>

  return {
    stream,
    push: (value: T) => {
      if (done) return
      queue.push(Promise.resolve(value))
    },
    done: () => {
      if (done) return
      signal = true
    },
  }
}

/**
 * Will automatically clamp to `-max -> max` if `min` is not provided
 */
export function clamp(toClamp: number, max: number, min?: number) {
  return Math.max(Math.min(toClamp, max), min ?? -max)
}

export function now() {
  return new Date().toISOString()
}
