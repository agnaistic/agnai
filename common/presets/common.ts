import { neat } from '../util'

export const gaslights = {
  alpaca: neat`
  Below is an instruction that describes a task. Write a response that appropriately completes the request.
  
  Write {{char}}'s next reply in a fictional roleplay chat between {{#each bot}}{{.name}}, {{/each}}{{char}}.
  
  {{char}}'s Persona: {{personality}}
  
  This scenario of the conversation: {{scenario}}
  
  This is how {{char}} should talk: {{example_dialogue}}
  
  Then the roleplay chat between {{#each bot}}{{.name}}, {{/each}}{{char}} begins.
  
  {{#each msg}}{{#if .isbot}}### Response:\n{{.name}}: {{.msg}}{{/if}}{{#if .isuser}}### Instruction:\n{{.name}}: {{.msg}}{{/if}}
  {{/each}}
  {{ujb}}
  ### Response:
  {{post}}
  `,
}
