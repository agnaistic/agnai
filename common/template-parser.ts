import { formatCharacter } from './characters'
import { grammar } from './grammar'
import { PromptParts, fillPromptWithLines } from './prompt'
import { AppSchema, Memory } from '/common/types'
import peggy from 'peggy'
import { elapsedSince } from './util'
import { Encoder } from './tokenize'
import { v4 } from 'uuid'

const parser = peggy.generate(grammar.trim(), {
  error: (stage, msg, loc) => {
    console.error({ loc, stage }, msg)
  },
})

type PNode = PlaceHolder | ConditionNode | IteratorNode | string

type PlaceHolder = { kind: 'placeholder'; value: Holder; values?: any; pipes?: string[] }
type ConditionNode = { kind: 'if'; value: Holder; values?: any; children: PNode[] }
type IteratorNode = { kind: 'each'; value: IterableHolder; children: CNode[] }

type CNode =
  | Exclude<PNode, { kind: 'each' }>
  | { kind: 'bot-prop'; prop: BotsProp }
  | { kind: 'history-prop'; prop: HistoryProp }
  | { kind: 'history-if'; prop: HistoryProp; children: CNode[] }
  | { kind: 'bot-if'; prop: BotsProp; children: CNode[] }

type Holder =
  | 'char'
  | 'user'
  | 'scenario'
  | 'personality'
  | 'example_dialogue'
  | 'history'
  | 'ujb'
  | 'post'
  | 'memory'
  | 'chat_age'
  | 'idle_duration'
  | 'all_personalities'
  | 'chat_embed'
  | 'user_embed'
  | 'impersonating'
  | 'system_prompt'
  | 'random'
  | 'roll'

type RepeatableHolder = Extract<
  Holder,
  'char' | 'user' | 'chat_age' | 'roll' | 'random' | 'idle_duration'
>

const repeatableHolders = new Set<RepeatableHolder>([
  'char',
  'user',
  'chat_age',
  'idle_duration',
  'random',
  'roll',
])

type IterableHolder = 'history' | 'bots'

type HistoryProp = 'i' | 'message' | 'dialogue' | 'name' | 'isuser' | 'isbot'
type BotsProp = 'i' | 'personality' | 'name'

export type TemplateOpts = {
  parts: Partial<PromptParts>
  chat: AppSchema.Chat

  char: AppSchema.Character
  replyAs: AppSchema.Character
  impersonate?: AppSchema.Character
  sender: AppSchema.Profile

  lines: string[]
  characters: Record<string, AppSchema.Character>
  lastMessage?: string

  chatEmbed?: Memory.UserEmbed<{ name: string }>[]
  userEmbed?: Memory.UserEmbed[]

  /** If present, history will be rendered last */
  limit?: {
    context: number
    encoder: Encoder
    output?: Record<string, string[]>
  }

  /**
   * Only allow repeatable placeholders. Excludes iterators, conditions, and prompt parts.
   */
  repeatable?: boolean
}

export function parseTemplate(template: string, opts: TemplateOpts) {
  if (opts.limit) {
    opts.limit.output = {}
  }

  if (opts.parts.systemPrompt) {
    opts.parts.systemPrompt = render(opts.parts.systemPrompt, opts)
  }

  if (opts.parts.ujb) {
    opts.parts.ujb = render(opts.parts.ujb, opts)
  }

  let output = render(template, opts)

  if (opts.limit && opts.limit.output) {
    for (const [id, lines] of Object.entries(opts.limit.output)) {
      const trimmed = fillPromptWithLines(
        opts.limit.encoder,
        opts.limit.context,
        output,
        lines
      ).reverse()
      output = output.replace(id, trimmed.join('\n'))
    }
  }

  return render(output, opts)
}

function render(template: string, opts: TemplateOpts) {
  try {
    const ast = parser.parse(template, {}) as PNode[]
    const output: string[] = []
    for (const parent of ast) {
      const result = renderNode(parent, opts)
      if (result) output.push(result)
    }
    return output.join('')
  } catch (err) {
    console.error({ err }, 'Failed to parse')
    throw err
  }
}

function renderNode(node: PNode, opts: TemplateOpts) {
  if (typeof node === 'string') {
    return node
  }

  switch (node.kind) {
    case 'placeholder': {
      return getPlaceholder(node, opts)
    }

    case 'each':
      return renderIterator(node.value, node.children, opts)

    case 'if':
      return renderCondition(node, node.children, opts)
  }
}

function renderProp(node: CNode, opts: TemplateOpts, entity: unknown, i: number) {
  if (typeof node === 'string') return node

  switch (node.kind) {
    case 'bot-if':
    case 'bot-prop': {
      const bot = entity as AppSchema.Character
      switch (node.prop) {
        case 'i':
          return i.toString()

        case 'name':
          return bot.name

        case 'personality':
          return formatCharacter(
            bot.name,
            bot.persona,
            bot.persona.kind /* || opts.chat.overrides.kind */ // looks like the || operator's left hand side is always truthy - @malfoyslastname
          )
      }
    }

    case 'history-if':
    case 'history-prop': {
      const line = entity as string
      switch (node.prop) {
        case 'i': {
          return i.toString()
        }

        case 'message': {
          const index = line.indexOf(':')
          return line.slice(index + 1).trim()
        }

        case 'name': {
          const index = line.indexOf(':')
          return line.slice(0, index)
        }

        case 'dialogue': {
          const index = line.indexOf(':')
          return line.slice(index + 1).trim()
        }

        case 'isbot':
        case 'isuser': {
          const index = line.indexOf(':')
          const name = line.slice(0, index)
          const sender = opts.impersonate?.name ?? opts.sender?.handle
          const match = name === sender
          return node.prop === 'isuser' ? match : !match
        }
      }
    }
  }
}

function renderCondition(node: ConditionNode, children: PNode[], opts: TemplateOpts) {
  if (opts.repeatable) return ''

  const value = getPlaceholder(node, opts)
  if (!value) return

  const output: string[] = []
  for (const child of children) {
    const result = renderNode(child, opts)
    if (result) output.push(result)
  }

  return output.join('')
}

function renderIterator(holder: IterableHolder, children: CNode[], opts: TemplateOpts) {
  if (opts.repeatable) return ''

  const output: string[] = []

  const entities =
    holder === 'bots'
      ? Object.values(opts.characters).filter((b) => !!b && b._id !== opts.replyAs._id)
      : opts.lines

  let i = 0
  for (const entity of entities) {
    let curr = ''
    for (const child of children) {
      if (typeof child === 'string') {
        curr += child
        continue
      }

      switch (child.kind) {
        case 'if':
        case 'placeholder': {
          const result = renderNode(child, opts)
          if (result) curr += result
          break
        }

        case 'bot-prop':
        case 'history-prop': {
          const result = renderProp(child, opts, entity, i)
          if (result) curr += result
          break
        }

        case 'bot-if':
        case 'history-if': {
          const prop = renderProp(child, opts, entity, i)
          if (!prop) break
          const result = renderEntityCondition(child.children, opts, entity, i)
          curr += result
          break
        }
      }
    }
    if (curr) output.push(curr)
    i++
  }

  if (holder === 'history' && opts.limit) {
    const id = '__' + v4() + '__'
    opts.limit.output![id] = output
    return id
  }

  return output.join('\n')
}

function renderEntityCondition(nodes: CNode[], opts: TemplateOpts, entity: unknown, i: number) {
  let result = ''

  for (const node of nodes) {
    const res = renderProp(node, opts, entity, i)
    if (res) result += res.toString()
  }

  return result
}

function getPlaceholder(node: PlaceHolder | ConditionNode, opts: TemplateOpts) {
  if (opts.repeatable && !repeatableHolders.has(node.value as any)) return ''

  switch (node.value) {
    case 'char':
      return opts.replyAs.name

    case 'user':
      return opts.impersonate?.name || opts.sender?.handle || 'You'

    case 'example_dialogue':
      return opts.parts.sampleChat?.join('\n') || ''

    case 'scenario':
      return opts.parts.scenario || opts.chat.scenario || opts.char.scenario

    case 'memory':
      return opts.parts.memory || ''

    case 'impersonating':
      return opts.parts.impersonality || ''

    case 'personality':
      return opts.parts.persona

    case 'ujb':
      return opts.parts.ujb || ''

    case 'post':
      return opts.parts.post?.join('\n') || ''

    case 'history': {
      if (opts.limit) {
        const id = `__${v4()}__`
        opts.limit.output![id] = opts.lines
        return id
      }

      return opts.lines.join('\n')
    }

    case 'chat_age':
      return elapsedSince(opts.chat.createdAt)

    case 'idle_duration':
      return lastMessage(opts.lastMessage || '')

    case 'all_personalities':
      return opts.parts.allPersonas?.join('\n') || ''

    case 'chat_embed':
      return opts.parts.chatEmbeds?.join('\n') || ''

    case 'user_embed':
      return opts.parts.userEmbeds?.join('\n') || ''

    case 'system_prompt':
      return opts.parts.systemPrompt || ''

    case 'random': {
      const values = node.values as string[]
      const rand = Math.random() * values.length
      return values[Math.floor(rand)]
    }

    case 'roll': {
      const max = +node.values
      const rand = Math.ceil(Math.random() * max)
      return rand.toString()
    }
  }
}

function lastMessage(value: string) {
  if (!value) return 'unknown'

  const date = new Date(value)
  if (isNaN(date.valueOf())) return 'unknown'
  return elapsedSince(date)
}
