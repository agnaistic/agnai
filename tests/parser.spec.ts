import { expect } from 'chai'
import './init'
import { reset, setRand, toBotMsg, toChar, toChat, toProfile, toUser, toUserMsg } from './util'
import { getPromptParts } from '/common/prompt'
import { TemplateOpts, parseTemplate } from '/common/template-parser'
import { AppSchema } from '/common/types'
import { getTokenCounter } from '/srv/tokenize'

const chars = [toChar('Robot'), toChar('Otherbot'), toChar('Thirdbot')]
const char = chars[0]
const { user } = toUser('sender')
const characters = chars.reduce<Record<string, AppSchema.Character>>(
  (prev, curr) => Object.assign(prev, { [curr._id]: curr }),
  {}
)
const chat = toChat(char)
const profile = toProfile('Sender')
const members = [profile]
const history = [
  toBotMsg(char, 'First message'),
  toUserMsg(profile, 'User response'),
  toBotMsg(chars[1], 'Multibot message'),
]
const lines = history.map(
  (h) => `${h.characterId ? characters[h.characterId]?.name : profile.handle}: ${h.msg}`
)

describe('Template parser tests', () => {
  before(reset)

  it('will render a basic template', () => {
    const actual = test(`Scenario: {{scenario}}\nPersona: {{personality}}`)
    expect(actual).toMatchSnapshot()
  })

  it('will render conditional', () => {
    const actual = test(`{{#if scenario}}Scenario: {{scenario}}{{/if}}`)
    expect(actual).toMatchSnapshot()
  })

  it('will not render missing conditional', () => {
    const actual = test(
      `{{#if scenario}}Scenario: {{scenario}}{{/if}}Persona: {{persona}}`,
      { char: toChar('Name', { scenario: '' }) },
      { scenario: '' }
    )
    expect(actual).toMatchSnapshot()
  })

  it('will iterate over messages', () => {
    const actual = test(`Scenario: {{scenario}}
{{#each history}}{{.i}} {{.dialogue}}{{/each}}`)
    expect(actual).toMatchSnapshot()
  })

  it('will create a full template', () => {
    const actual = test(
      `
Scenario: {{scenario}}
{{char}}'s persona: {{persona}}
{{#each bots}}{{.name}}'s persona: {{.persona}}{{/each}}

{{#if example_dialogue}}Example dialogue: {{example_dialogue}}{{/if}}
<START>
{{#each history}}- {{.name}}: {{.dialogue}}{{/each}}
(System: Reply as {{char}})
{{post}}`.trim()
    )
    expect(actual).toMatchSnapshot()
  })

  it('will handle history conditions correctly', () => {
    const actual = test(
      `
Scenario: {{scenario}}
{{char}}'s persona: {{persona}}
{{#each bots}}{{.name}}'s persona: {{.persona}}{{/each}}

{{#each history}}{{#if .isbot}}Assistant: {{/if}}{{#if .isuser}}Human: {{/if}}{{.name}}: {{.dialogue}}{{/each}}
{{post}}`.trim()
    )
    expect(actual).toMatchSnapshot()
  })

  it('will render a random placeholder', () => {
    {
      setRand(1 / 3)
      const actual = test(`Random {{random: one, two, three}}`)
      expect(actual).to.equal('Random two')
    }

    {
      setRand(2 / 3)
      const actual = test(`Random {{random: one, two, three}}`)
      expect(actual).to.equal('Random three')
    }
  })
})

function test(
  template: string,
  overrides: Partial<TemplateOpts> = {},
  charOverrides: Partial<AppSchema.Character> = {}
) {
  return parseTemplate(template, getParseOpts(overrides, charOverrides)).parsed
}

function getParseOpts(
  overrides: Partial<TemplateOpts> = {},
  charOverrides: Partial<AppSchema.Character> = {}
) {
  const overChat = overrides.char ? toChat(overrides.char) : chat
  const overChar = { ...char, ...charOverrides }
  const parts = getPromptParts(
    {
      char: overChar,
      characters,
      chat: overChat,
      members,
      replyAs: overChar,
      user,
      kind: 'send',
      sender: profile,
      chatEmbeds: [],
      userEmbeds: [],
    },
    lines,
    getTokenCounter('openai', 'turbo')
  )

  const base: TemplateOpts = {
    char,
    characters,
    chat: overChat,
    replyAs: char,
    lines,
    parts,
    sender: profile,
    lastMessage: '',
    ...overrides,
  }

  return base
}
