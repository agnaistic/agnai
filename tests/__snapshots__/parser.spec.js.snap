// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template parser tests will create a full template 1`] = `
"Scenario: SCENARIO
Robot's persona: PERSONA
Otherbot's persona: PERSONA
Thirdbot's persona: PERSONA

Example dialogue: Robot: SAMPLE_CHAT
<START>
- Robot: First message
- Sender: User response
- Otherbot: Multibot message
(System: Reply as Robot)
Robot:"
`;

exports[`Template parser tests will handle history conditions correctly 1`] = `
"Scenario: SCENARIO
Robot's persona: PERSONA
Otherbot's persona: PERSONA
Thirdbot's persona: PERSONA

Assistant: Robot: First message
Human: Sender: User response
Assistant: Otherbot: Multibot message
Robot:"
`;

exports[`Template parser tests will iterate over messages 1`] = `
"Scenario: SCENARIO
0 First message
1 User response
2 Multibot message"
`;

exports[`Template parser tests will not render missing conditional 1`] = `"Persona: PERSONA"`;

exports[`Template parser tests will render a basic template 1`] = `
"Scenario: SCENARIO
Persona: PERSONA"
`;

exports[`Template parser tests will render conditional 1`] = `"Scenario: SCENARIO"`;
